File: packages/engine/src/board.ts
/**
 * SPEC §1–§3, §5 — board encoding, line merge, moves and spawning.
 * A board is a Uint8Array(16) of exponents, row-major.
 */
import type { Rng } from './rng.ts';

export type Board = Uint8Array;

export const UP = 0;
export const DOWN = 1;
export const LEFT = 2;
export const RIGHT = 3;
export type Direction = 0 | 1 | 2 | 3;
export const DIRECTIONS: readonly Direction[] = [UP, DOWN, LEFT, RIGHT];
export const DIRECTION_LETTERS = 'UDLR';
export const DIRECTION_NAMES = ['up', 'down', 'left', 'right'] as const;
export type DirectionName = (typeof DIRECTION_NAMES)[number];

/** LINES[dir][line] = the four cell indices, starting at the edge tiles slide toward. */
export const LINES: readonly (readonly (readonly number[])[])[] = [
  [0, 1, 2, 3].map((c) => [c, 4 + c, 8 + c, 12 + c]),
  [0, 1, 2, 3].map((c) => [12 + c, 8 + c, 4 + c, c]),
  [0, 1, 2, 3].map((r) => [4 * r, 4 * r + 1, 4 * r + 2, 4 * r + 3]),
  [0, 1, 2, 3].map((r) => [4 * r + 3, 4 * r + 2, 4 * r + 1, 4 * r]),
];

const HEX = '0123456789abcdefghijklmnopqrstuvwxyz';

export function emptyBoard(): Board {
  return new Uint8Array(16);
}

export function boardToHex(board: Board): string {
  let s = '';
  for (let i = 0; i < 16; i++) s += HEX[board[i]];
  return s;
}

export function boardFromHex(hex: string): Board {
  if (hex.length !== 16) throw new Error(`boardHex must be 16 chars, got ${hex.length}`);
  const b = new Uint8Array(16);
  for (let i = 0; i < 16; i++) {
    const v = HEX.indexOf(hex[i].toLowerCase());
    if (v < 0) throw new Error(`invalid boardHex character '${hex[i]}'`);
    b[i] = v;
  }
  return b;
}

/** Board as tile values (0 for empty) in a 4×4 matrix — convenient for agents and APIs. */
export function boardToMatrix(board: Board): number[][] {
  const m: number[][] = [];
  for (let r = 0; r < 4; r++) {
    const row: number[] = [];
    for (let c = 0; c < 4; c++) {
      const e = board[r * 4 + c];
      row.push(e === 0 ? 0 : 2 ** e);
    }
    m.push(row);
  }
  return m;
}

export function boardsEqual(a: Board, b: Board): boolean {
  for (let i = 0; i < 16; i++) if (a[i] !== b[i]) return false;
  return true;
}

export function maxExponent(board: Board): number {
  let m = 0;
  for (let i = 0; i < 16; i++) if (board[i] > m) m = board[i];
  return m;
}

export function maxTile(board: Board): number {
  const e = maxExponent(board);
  return e === 0 ? 0 : 2 ** e;
}

export function emptyCount(board: Board): number {
  let n = 0;
  for (let i = 0; i < 16; i++) if (board[i] === 0) n++;
  return n;
}

const lineBuf = new Uint8Array(4);
const resBuf = new Uint8Array(4);

/**
 * Apply `dir` to `board` in place. Returns the score gained, or -1 if the
 * move is invalid (in which case the board is untouched).
 */
export function moveInPlace(board: Board, dir: Direction): number {
  const lines = LINES[dir];
  let gained = 0;
  let changed = false;
  for (let l = 0; l < 4; l++) {
    const idx = lines[l];
    let n = 0;
    for (let k = 0; k < 4; k++) {
      const v = board[idx[k]];
      if (v !== 0) lineBuf[n++] = v;
    }
    resBuf.fill(0);
    let out = 0;
    let i = 0;
    while (i < n) {
      if (i + 1 < n && lineBuf[i] === lineBuf[i + 1]) {
        const e = lineBuf[i] + 1;
        resBuf[out++] = e;
        gained += 2 ** e;
        i += 2;
      } else {
        resBuf[out++] = lineBuf[i];
        i += 1;
      }
    }
    // Lines are independent, so an unchanged line is simply left alone.
    for (let k = 0; k < 4; k++) {
      if (board[idx[k]] !== resBuf[k]) {
        board[idx[k]] = resBuf[k];
        changed = true;
      }
    }
  }
  return changed ? gained : -1;
}

export interface MoveResult {
  board: Board;
  gained: number;
  changed: boolean;
}

export function move(board: Board, dir: Direction): MoveResult {
  const b = board.slice();
  const g = moveInPlace(b, dir);
  return g < 0 ? { board, gained: 0, changed: false } : { board: b, gained: g, changed: true };
}

export function canMove(board: Board, dir: Direction): boolean {
  for (const idx of LINES[dir]) {
    for (let k = 1; k < 4; k++) {
      const prev = board[idx[k - 1]];
      const cur = board[idx[k]];
      if (cur !== 0 && (prev === 0 || prev === cur)) return true;
    }
  }
  return false;
}

export function validMoves(board: Board): Direction[] {
  return DIRECTIONS.filter((d) => canMove(board, d));
}

export function isOver(board: Board): boolean {
  for (let d = 0; d < 4; d++) if (canMove(board, d as Direction)) return false;
  return true;
}

export interface Spawn {
  index: number;
  exponent: number;
}

/** SPEC §5. Returns the spawned tile, or null if the board is full. */
export function spawn(board: Board, rng: Rng): Spawn | null {
  const empties: number[] = [];
  for (let i = 0; i < 16; i++) if (board[i] === 0) empties.push(i);
  if (empties.length === 0) return null;
  const index = empties[rng.below(empties.length)];
  const exponent = rng.below(10) === 0 ? 2 : 1;
  board[index] = exponent;
  return { index, exponent };
}

/** Per-tile motion for UI animation: where each source tile ends up. */
export interface TileMotion {
  from: number;
  to: number;
  exponent: number;
  /** true for both tiles that combine into a merged tile */
  merged: boolean;
}

export function moveMotions(board: Board, dir: Direction): TileMotion[] {
  const motions: TileMotion[] = [];
  for (const idx of LINES[dir]) {
    const tiles: number[] = [];
    for (let k = 0; k < 4; k++) if (board[idx[k]] !== 0) tiles.push(idx[k]);
    let out = 0;
    let i = 0;
    while (i < tiles.length) {
      const a = tiles[i];
      if (i + 1 < tiles.length && board[a] === board[tiles[i + 1]]) {
        motions.push({ from: a, to: idx[out], exponent: board[a], merged: true });
        motions.push({ from: tiles[i + 1], to: idx[out], exponent: board[a], merged: true });
        i += 2;
      } else {
        motions.push({ from: a, to: idx[out], exponent: board[a], merged: false });
        i += 1;
      }
      out++;
    }
  }
  return motions;
}

export function parseDirection(input: string | number): Direction {
  if (typeof input === 'number') {
    if (input === 0 || input === 1 || input === 2 || input === 3) return input;
    throw new Error(`invalid direction ${input}`);
  }
  const s = input.trim().toLowerCase();
  const byName = DIRECTION_NAMES.indexOf(s as DirectionName);
  if (byName >= 0) return byName as Direction;
  const byLetter = DIRECTION_LETTERS.indexOf(s.toUpperCase());
  if (s.length === 1 && byLetter >= 0) return byLetter as Direction;
  throw new Error(`invalid direction '${input}'`);
}

File: engines/java/src/main/java/dev/g2048/engine/Board.java
package dev.g2048.engine;

/**
 * Board helpers over {@code byte[16]} exponent arrays, row-major (SPEC §1–§5).
 */
public final class Board {
  private Board() {}

  public static final int UP = 0, DOWN = 1, LEFT = 2, RIGHT = 3;
  public static final String DIRECTION_LETTERS = "UDLR";
  public static final String[] DIRECTION_NAMES = {"up", "down", "left", "right"};
  private static final String HEX_DIGITS = "0123456789abcdefghijklmnopqrstuvwxyz";

  /** LINES[dir][line][k] = cell index, starting at the edge tiles slide toward (SPEC §3). */
  public static final int[][][] LINES = new int[4][4][4];

  static {
    for (int k = 0; k < 4; k++) {
      for (int j = 0; j < 4; j++) {
        LINES[UP][k][j] = j * 4 + k;
        LINES[DOWN][k][j] = (3 - j) * 4 + k;
        LINES[LEFT][k][j] = k * 4 + j;
        LINES[RIGHT][k][j] = k * 4 + 3 - j;
      }
    }
  }

  public static String hex(byte[] b) {
    char[] s = new char[16];
    for (int i = 0; i < 16; i++) s[i] = HEX_DIGITS.charAt(b[i]);
    return new String(s);
  }

  public static byte[] fromHex(String hex) {
    if (hex.length() != 16) throw new IllegalArgumentException("boardHex must be 16 chars, got " + hex.length());
    byte[] b = new byte[16];
    for (int i = 0; i < 16; i++) {
      int v = HEX_DIGITS.indexOf(Character.toLowerCase(hex.charAt(i)));
      if (v < 0) throw new IllegalArgumentException("invalid boardHex character '" + hex.charAt(i) + "'");
      b[i] = (byte) v;
    }
    return b;
  }

  public static int maxExponent(byte[] b) {
    int m = 0;
    for (byte e : b) if (e > m) m = e;
    return m;
  }

  /** Largest tile value (0 for an empty board). */
  public static long maxTile(byte[] b) {
    int e = maxExponent(b);
    return e == 0 ? 0 : 1L << e;
  }

  /** Applies dir in place. Returns the score gained, or -1 if invalid (board untouched). */
  public static long moveInPlace(byte[] b, int dir) {
    long gained = 0;
    boolean changed = false;
    int[][] lines = LINES[dir];
    byte[] tiles = new byte[4];
    byte[] res = new byte[4];
    for (int l = 0; l < 4; l++) {
      int[] idx = lines[l];
      int n = 0;
      for (int k = 0; k < 4; k++) {
        byte v = b[idx[k]];
        if (v != 0) tiles[n++] = v;
      }
      int out = 0;
      for (int i = 0; i < n; ) {
        if (i + 1 < n && tiles[i] == tiles[i + 1]) {
          int e = tiles[i] + 1;
          res[out] = (byte) e;
          gained += 1L << e;
          i += 2;
        } else {
          res[out] = tiles[i];
          i++;
        }
        out++;
      }
      for (; out < 4; out++) res[out] = 0;
      for (int k = 0; k < 4; k++) {
        if (b[idx[k]] != res[k]) {
          b[idx[k]] = res[k];
          changed = true;
        }
      }
    }
    return changed ? gained : -1;
  }

  /** Result of a non-mutating move. */
  public record MoveResult(byte[] board, long gained, boolean changed) {}

  public static MoveResult move(byte[] b, int dir) {
    byte[] nb = b.clone();
    long g = moveInPlace(nb, dir);
    if (g < 0) return new MoveResult(b.clone(), 0, false);
    return new MoveResult(nb, g, true);
  }

  public static boolean canMove(byte[] b, int dir) {
    int[][] lines = LINES[dir];
    for (int l = 0; l < 4; l++) {
      int[] idx = lines[l];
      for (int k = 1; k < 4; k++) {
        byte prev = b[idx[k - 1]], cur = b[idx[k]];
        if (cur != 0 && (prev == 0 || prev == cur)) return true;
      }
    }
    return false;
  }

  public static int[] validMoves(byte[] b) {
    int[] tmp = new int[4];
    int n = 0;
    for (int d = 0; d < 4; d++) if (canMove(b, d)) tmp[n++] = d;
    return java.util.Arrays.copyOf(tmp, n);
  }

  public static boolean isOver(byte[] b) {
    for (int d = 0; d < 4; d++) if (canMove(b, d)) return false;
    return true;
  }

  /** Spawn result; index -1 means the board was full (no draws). */
  public record Spawned(int index, int exponent) {
    public boolean ok() {
      return index >= 0;
    }
  }

  /** Places a tile per SPEC §5. Returns packed (index << 8 | exponent), or -1 if full. */
  public static int spawnPacked(byte[] b, Rng r) {
    int[] empties = new int[16];
    int n = 0;
    for (int i = 0; i < 16; i++) if (b[i] == 0) empties[n++] = i;
    if (n == 0) return -1;
    int index = empties[(int) r.below(n)];
    int exponent = r.below(10) == 0 ? 2 : 1;
    b[index] = (byte) exponent;
    return index << 8 | exponent;
  }

  public static Spawned spawn(byte[] b, Rng r) {
    int p = spawnPacked(b, r);
    return p < 0 ? new Spawned(-1, 0) : new Spawned(p >>> 8, p & 0xff);
  }

  /** Accepts names ("left"), letters ("L") or digits ("2"). */
  public static int parseDirection(String s) {
    String t = s.trim().toLowerCase();
    for (int i = 0; i < 4; i++) if (t.equals(DIRECTION_NAMES[i])) return i;
    if (t.length() == 1) {
      int i = "udlr".indexOf(t.charAt(0));
      if (i >= 0) return i;
      char c = t.charAt(0);
      if (c >= '0' && c <= '3') return c - '0';
    }
    throw new IllegalArgumentException("invalid direction '" + s + "'");
  }
}
